src/lib/verifier/prod.ml

```diff
@@ -354,9 +356,17 @@ let with_retry ~logger f =
 
 let verify_blockchain_snarks {worker; logger} chains =
   with_retry ~logger (fun () ->
+      [%log debug] "Before wait for the verifier process" ;
       let%bind {connection; _} = !worker in
-      Worker.Connection.run connection ~f:Worker.functions.verify_blockchains
-        ~arg:chains )
+      [%log debug] "After wait for the verifier process" ;
```

Comment:
Looking at this again, I've realised there's more going on here than I originally thought. The worker doesn't get re-initialised until the first request for verification here, which clearly seems wrong to me.

Can you change the code around `new_worker` above to be roughly
```ocaml
        let new_worker = Ivar.create () in
        worker_ref := new_worker;
        don't_wait_for (
          let%bind exit_metadata =
            match%map wait_safe process with
            | Ok res ->
                [ ( "exit_status"
                  , `String (Unix.Exit_or_signal.to_string_hum res) ) ]
            | Error err ->
                [ ("exit_status", `String "Unknown: wait threw an error")
                ; ("exn", Error_json.error_to_yojson err) ]
          in
          [%log info] "verifier successfully stopped"
            ~metadata:
              ( ("verifier_pid", `Int (Process.pid process |> Pid.to_int))
              :: exit_metadata ) ;
          Child_processes.Termination.remove pids pid ;
          let%map worker = create_worker () in
          on_worker worker ;
          Ivar.fill new_worker worker )
```
which will make it restart immediately after being killed. Then -- the original reason I was going to comment -- the noisy logs here can only be printed when they are relevant:
```ocaml
  let%bind {connection; _} =
    let ivar = !worker in
    match Ivar.peek ivar with
    | Some worker ->
        Deferred.return worker (* The worker is ready, use it directly *)
    | None ->
        [%log debug] "Waiting for the verifier process to restart";
        let%map worker = Ivar.read ivar in
        [%log debug] "Verifier process has restarted; finished waiting";
        worker
  in
```

You'll also need to tweak the type of `worker` to be an `_ Ivar.t ref`, and change the original assignment to be a `worker_ref := Ivar.create_full worker`, but this should be trivial.